src/CaptureClient/CaptureEventProcessorTest.cpp

```diff
@@ -1170,6 +1181,110 @@ TEST(CaptureEventProcessor, CanHandleThreadStateSlices) {
+
+  constexpr uint64_t kCallstackId = 24;
+
+  ClientCaptureEvent thread_state_slice_without_known_callstack_id_event;
+  ThreadStateSlice* thread_state_slice_without_known_callstack_id =
+      thread_state_slice_without_known_callstack_id_event.mutable_thread_state_slice();
+  thread_state_slice_without_known_callstack_id->set_duration_ns(100);
+  thread_state_slice_without_known_callstack_id->set_end_timestamp_ns(200);
+  thread_state_slice_without_known_callstack_id->set_pid(14);
+  thread_state_slice_without_known_callstack_id->set_tid(24);
+  thread_state_slice_without_known_callstack_id->set_thread_state(ThreadStateSlice::kRunnable);
+  thread_state_slice_without_known_callstack_id->set_switch_out_or_wakeup_callstack_status(
+      ThreadStateSlice::kCallstackSet);
+  thread_state_slice_without_known_callstack_id->set_switch_out_or_wakeup_callstack_id(
+      kCallstackId);
+
+  ClientCaptureEvent thread_state_slice_waiting_for_callstack_event;
+  ThreadStateSlice* thread_state_slice_waiting_for_callstack =
+      thread_state_slice_waiting_for_callstack_event.mutable_thread_state_slice();
+  thread_state_slice_waiting_for_callstack->set_duration_ns(100);
+  thread_state_slice_waiting_for_callstack->set_end_timestamp_ns(200);
+  thread_state_slice_waiting_for_callstack->set_pid(14);
+  thread_state_slice_waiting_for_callstack->set_tid(24);
+  thread_state_slice_waiting_for_callstack->set_thread_state(ThreadStateSlice::kInterruptibleSleep);
+  thread_state_slice_waiting_for_callstack->set_switch_out_or_wakeup_callstack_status(
+      ThreadStateSlice::kWaitingForCallstack);
+  thread_state_slice_waiting_for_callstack->set_switch_out_or_wakeup_callstack_id(0);
+
+  EXPECT_CALL(listener, OnUniqueCallstack).Times(0);
+  EXPECT_CALL(listener, OnThreadStateSlice).Times(0);
+
+  EXPECT_DEATH(event_processor->ProcessEvent(thread_state_slice_without_known_callstack_id_event),
+               "");
+  EXPECT_DEATH(event_processor->ProcessEvent(thread_state_slice_waiting_for_callstack_event), "");
+}
+
+TEST(CaptureEventProcessor, CanHandleThreadStateSlicesWithCallstacks) {
+  MockCaptureListener listener;
+  auto event_processor =
+      CaptureEventProcessor::CreateForCaptureListener(&listener, std::filesystem::path{}, {});
+
+  constexpr uint64_t kCallstackId = 24;
+  constexpr uint64_t kFrame1 = 1;
+  constexpr uint64_t kFrame2 = 2;
+  constexpr uint64_t kFrame3 = 3;
+
+  ClientCaptureEvent interned_callstack_event;
+  InternedCallstack* interned_callstack = interned_callstack_event.mutable_interned_callstack();
+  interned_callstack->set_key(kCallstackId);
+  interned_callstack->mutable_intern()->add_pcs(kFrame1);
+  interned_callstack->mutable_intern()->add_pcs(kFrame2);
+  interned_callstack->mutable_intern()->add_pcs(kFrame3);
+  interned_callstack->mutable_intern()->set_type(orbit_grpc_protos::Callstack::kComplete);
+
+  ClientCaptureEvent runnable_event;
```

Comment:
Name is a bit strange: `runnable_thread_state_slice_event` is more fitting with the naming pattern.